bdai_ros2_wrappers/bdai_ros2_wrappers/executors.py

```diff
@@ -0,0 +1,395 @@
+                # careful with back references to self!
+                worker = threading.Thread(target=AutoScalingThreadPool._do_work, args=(weak_self, False), daemon=True)
+                worker.start()
+                self._workers.add(worker)
+            # register workers for external joining on interpreter shutdown
+            AutoScalingThreadPool._all_workers.update(self._workers)
+
+        self._waitqueues: typing.Dict[typing.Callable[..., typing.Any], collections.deque] = collections.defaultdict(
+            collections.deque
+        )
+        self._runlists: typing.Dict[typing.Callable[..., typing.Any], typing.List[AutoScalingThreadPool.Work]] = (
+            collections.defaultdict(list)
+        )
+        self._runslots = threading.Semaphore(0)
+
+    @property
+    def workers(self) -> typing.List[threading.Thread]:
+        with self._submit_lock:
+            # avoid races and proactively drop shutdown workers
+            self._workers = weakref.WeakSet([w for w in self._workers if w.is_alive()])
+            return list(self._workers)
+
+    @property
+    def working(self) -> bool:
+        with self._submit_lock:
+            return len(self._runlists) > 0 or len(self._waitqueues) > 0
+
+    @property
+    def capped(self) -> bool:
+        with self._submit_lock:
+            return len(self._waitqueues) > 0
+
+    def wait(self, timeout: typing.Optional[float] = None) -> bool:
+        with self._submit_lock:
+            futures = [work.future for runlist in self._runlists.values() for work in runlist]
+            futures += [work.future for waitqueue in self._waitqueues.values() for work in waitqueue]
+        done, not_done = concurrent.futures.wait(futures, timeout=timeout)
+        return len(not_done) == 0
+
+    def _cleanup_after(self, work: "AutoScalingThreadPool.Work") -> bool:
+        complete = True
+        with self._submit_lock:
+            self._runlists[work.fn].remove(work)
+            if work.fn in self._waitqueues and self._waitqueues[work.fn]:  # continue with pending work
+                work = self._waitqueues[work.fn].popleft()
+                while work.cancelled() and self._waitqueues[work.fn]:
+                    work.notify_cancelation()
+                    work = self._waitqueues[work.fn].popleft()
+                if not self._waitqueues[work.fn]:
+                    del self._waitqueues[work.fn]
+                if not work.cancelled():
+                    self._runlists[work.fn].append(work)
+                    self._runqueue.put(work)  # actually submit work
+                    complete = False
+                else:
+                    work.notify_cancelation()
+            if not self._runlists[work.fn]:
+                del self._runlists[work.fn]
+        return complete
+
+    @classmethod
+    def _do_work(cls, executor_weakref: weakref.ref, stop_on_timeout: bool = True) -> None:
+        logger = logging.getLogger(__name__)
```

Comment:
How does this interact with RcUtilLogger and more specifically does it get captured by rosbags?

Reply:
It doesn't interact at all right now. We can make it so, however, by forwarding a logger instance on pool instantiation.

Reply:
Done in https://github.com/bdaiinstitute/ros_utilities/pull/22/commits/2ee1d41e5649c17cd8e75d8b4df02c4f70c99906.